File: app/src/main/java/sk/styk/martin/apkanalyzer/business/task/AppListLoader.java
package sk.styk.martin.apkanalyzer.business.task;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;

import java.util.List;

import sk.styk.martin.apkanalyzer.business.service.AppBasicDataService;
import sk.styk.martin.apkanalyzer.model.list.AppListData;

/**
 * Loader async task for items for list view on AppListFragment
 * <p>
 * Created by Martin Styk on 15.06.2017.
 */
public class AppListLoader extends ApkAnalyzerAbstractAsyncLoader<List<AppListData>> {
    public static final int ID = 1;

    private final PackageManager mPackageManager;
    private final AppBasicDataService installedAppsService;

    PackageIntentReceiver mPackageObserver;

    public AppListLoader(Context context) {
        super(context);

        mPackageManager = getContext().getPackageManager();
        installedAppsService = new AppBasicDataService(mPackageManager);
    }

    /**
     * This is where the bulk of our work is done.  This function is
     * called in a background thread and should generate a new set of
     * data to be published by the loader.
     */
    @Override
    public List<AppListData> loadInBackground() {
        return installedAppsService.getAll();
    }

    /**
     * Handles a request to start the Loader.
     */
    @Override
    protected void onStartLoading() {
        if (items != null) {
            // If we currently have a result available, deliver it
            // immediately.
            deliverResult(items);
        }

        // Start watching for changes in the app data.
        if (mPackageObserver == null) {
            mPackageObserver = new PackageIntentReceiver(this);
        }

        if (takeContentChanged() || items == null) {
            // If the data has changed since the last time it was loaded
            // or is not currently available, start a load.
            forceLoad();
        }
    }

    /**
     * Handles a request to stop the Loader.
     */
    @Override
    protected void onStopLoading() {
        // Attempt to cancel the current load task if possible.
        cancelLoad();
    }

    /**
     * Handles a request to completely reset the Loader.
     */
    @Override
    protected void onReset() {
        super.onReset();

        onStopLoading();

        if (items != null) {
            onReleaseResources(items);
            items = null;
        }

        // Stop monitoring for changes.
        if (mPackageObserver != null) {
            getContext().unregisterReceiver(mPackageObserver);
            mPackageObserver = null;
        }
    }


    /**
     * Helper class to look for interesting changes to the installed apps
     * so that the loader can be updated.
     */
    public static class PackageIntentReceiver extends BroadcastReceiver {
        final AppListLoader mLoader;

        public PackageIntentReceiver(AppListLoader loader) {
            mLoader = loader;
            IntentFilter filter = new IntentFilter(Intent.ACTION_PACKAGE_ADDED);
            filter.addAction(Intent.ACTION_PACKAGE_REMOVED);
            filter.addAction(Intent.ACTION_PACKAGE_CHANGED);
            filter.addDataScheme("package");
            mLoader.getContext().registerReceiver(this, filter);
            // Register for events related to sdcard installation.
            IntentFilter sdFilter = new IntentFilter();
            sdFilter.addAction(Intent.ACTION_EXTERNAL_APPLICATIONS_AVAILABLE);
            sdFilter.addAction(Intent.ACTION_EXTERNAL_APPLICATIONS_UNAVAILABLE);
            mLoader.getContext().registerReceiver(this, sdFilter);
        }

        @Override
        public void onReceive(Context context, Intent intent) {
            // Tell the loader about the change.
            mLoader.onContentChanged();
        }
    }

}

File: app/src/main/java/sk/styk/martin/apkanalyzer/business/analysis/task/AppListLoader.kt
package sk.styk.martin.apkanalyzer.business.analysis.task

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import sk.styk.martin.apkanalyzer.business.analysis.logic.launcher.AppBasicDataService
import sk.styk.martin.apkanalyzer.business.base.task.ApkAnalyzerAbstractAsyncLoader
import sk.styk.martin.apkanalyzer.model.list.AppListData

/**
 * Loader async task for items for list view on AppListFragment
 *
 * @author Martin Styk
 * @version 15.06.2017.
 */
class AppListLoader(context: Context) : ApkAnalyzerAbstractAsyncLoader<List<AppListData>>(context) {

    private val installedAppsService = AppBasicDataService(getContext().packageManager)

    private var mPackageObserver: PackageIntentReceiver? = null

    override fun loadInBackground(): List<AppListData> {
        return installedAppsService.getAll()
    }

    /**
     * Handles a request to start the Loader.
     */
    override fun onStartLoading() {
        // Start watching for changes in the app data.
        if (mPackageObserver == null) {
            mPackageObserver = PackageIntentReceiver(this)
        }

        super.onStartLoading()
    }

    override fun onReset() {
        // Stop monitoring for changes.
        if (mPackageObserver != null) {
            context.unregisterReceiver(mPackageObserver)
            mPackageObserver = null
        }

        super.onReset()
    }

    /**
     * Helper class to look for interesting changes to the installed apps
     * so that the loader can be updated.
     */
    class PackageIntentReceiver(private val mLoader: AppListLoader) : BroadcastReceiver() {

        init {
            val filter = IntentFilter(Intent.ACTION_PACKAGE_ADDED)
            filter.addAction(Intent.ACTION_PACKAGE_REMOVED)
            filter.addAction(Intent.ACTION_PACKAGE_CHANGED)
            filter.addDataScheme("package")
            mLoader.context.registerReceiver(this, filter)

            // Register for events related to sdcard installation.
            val sdFilter = IntentFilter()
            sdFilter.addAction(Intent.ACTION_EXTERNAL_APPLICATIONS_AVAILABLE)
            sdFilter.addAction(Intent.ACTION_EXTERNAL_APPLICATIONS_UNAVAILABLE)
            mLoader.context.registerReceiver(this, sdFilter)
        }

        override fun onReceive(context: Context, intent: Intent) {
            // Tell the loader about the change.
            mLoader.onContentChanged()
        }
    }

    companion object {
        const val ID = 1
    }
}
